### PR TITLE
Add a define to signify breaking API changes

### PR DESCRIPTION
**Pull request by waTeim**
A define that will allow embedding C and C++ programs to compile conditionally when non-backward compatible changes are introduced.

### CODE REVIEW

**Comment by yuyichao** [comment]
So the question is, how often do you expect this to change. Should it be bump whenever there's a breaking change on the master, or should it only be changed on released versions.

**Comment by waTeim** [comment]
Good question.   The answer is master. 

The practical issue is that dev can take a year to become release, and I understand the reluctance to have a 0.4.x where x changes before release  -- but that would also be a solution.  But people use dev before, in fact there are probably a bunch of packages that don't currently even work with 0.3 (e.g. Cxx).  Programs that embed must support dev as if it's released therefore.  Also, it's unrealistic to expect users to update their version of dev whenever this happens.  

Additionally, the Julia release keeps a lot of these details hidden from users and so releases are not even tied to the embed API especially when considering back-ported features.  And having the API version only updated when a release happens is also redundant since the information is already there (JULIA_VERSION_MINOR, etc).

Maybe having commit# of julia.h, better idea though I'm not sure how to automatically reconcile that with features.

The question is how to determine when the API number needs to be changed then?  I suggest it must change when the embedding test breaks (not created yet).

Currently there is an undetectable breaking change 

atexit_hook(void) was changed to atexit_hook(int), and there's not way to determine which one will be encountered form julia.h.  That's an emergency for me.

Thus this pull request.

**Comment by yuyichao** [comment]
I think that make sense. But the next question is, how should we make sure this is useful.

I know there's large projects that use this. I'm just wondering how can this be enforced or does it usually works well enough to just let every developers know about it?

**Comment by yuyichao** [comment]
An related question is, `libjulia` doesn't have any soname (yet). How do you want to handle this at ABI level.

This should probably also need some document (which should include the answers to my questions above, i.e. policies and how to use them)

**Comment by waTeim** [comment]
It's sort of working especially for big things, usually there's some other define that accompanies big changes that can be (has been) used as a proxy.  And usually big things are usually accompanied with a post to julia-dev.  That doesn't mean it's always like that (present example).  Either way there's usually catch up to play, and during that time things are broken.

As far as ABI goes, the library is currently found by julia using `@rpath`, and so is always in a known place relative to the julia that uses it and so up-till-now there has been no need for so versions, however this brings up a point.  Could this all be flattened, maybe it could.  I don't think there's an immediate problem, but this could be a stepping stone to that.  I had not thought about that.  I think that libjulia.so.x.y would be usable by julia that expects libjulia.x.z and the API version might be a source of x.  embedders could follow suit.

TL;DR people get busy, an automatic system would be best.

I think automatic tests.
A auto-generated email to a list would be great if possible.
